### PR TITLE
8307927: C2: "malformed control flow" with irreducible loop

### PR DESCRIPTION
**Pull request by rwestrel**
The test contains a loop nest with 2 loops. The outer loop is an
irreducible loop. The safepoint for that loop is also in the inner
loop. Because `IdealLoopTree::check_safepts()` ignores irreducible
loops, that safepoint is not marked as being required and is
eliminated from the inner loop. The inner loop is then optimized out
and the outer loop becomes an infinite loop with no safepoint (a
single node loop). That, in turn, causes the loop to be eliminated
because it has not use and the assert fires.

The fix I propose is to make `IdealLoopTree::check_safepts()` work
with irreducible loops. I think
`IdealLoopTree::allpaths_check_safepts()` can be used for that. When
working on this I wondered if that method could be called with a loop
whose head has more than 3 inputs. I couldn't write a test case with
an irreducible loop whose head had more than 3 inputs but I added an
assert in the method and ran some testing. That assert fired so I also
propose to tweak the method so it's robust in that case.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8307927](https://bugs.openjdk.org/browse/JDK-8307927): C2: "malformed control flow" with irreducible loop (**Bug** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [3bb72120](https://git.openjdk.org/jdk/pull/14522/files/3bb721201f9eb453e92aa58ff0ed814c25b1297b)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14522/head:pull/14522` \
`$ git checkout pull/14522`

Update a local copy of the PR: \
`$ git checkout pull/14522` \
`$ git pull https://git.openjdk.org/jdk.git pull/14522/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14522`

View PR using the GUI difftool: \
`$ git pr show -t 14522`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14522.diff">https://git.openjdk.org/jdk/pull/14522.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14522#issuecomment-1594932385)

### CODE REVIEW

**Review by TobiHartmann — APPROVED** [review]
Looks reasonable to me. All tests passed.

@eme64 Please have a look as well.

File: src/hotspot/share/opto/loopnode.cpp [hunk 1]
```diff
@@ -3514,9 +3514,9 @@ bool IdealLoopTree::beautify_loops( PhaseIdealLoop *phase ) {
 // encountered.  Helper for check_safepts.
 void IdealLoopTree::allpaths_check_safepts(VectorSet &visited, Node_List &stack) {
```
Comment (eme64) [inline src/hotspot/share/opto/loopnode.cpp:3515]: @rwestrel you should update the description here. Suggestion:
```
Allpaths backwards scan. Starting at the head, traversing all backedges, and the body. Terminating each path at first safepoint encountered.  Helper for check_safepts.
```

Comment (eme64) [reply src/hotspot/share/opto/loopnode.cpp:3515]: Also the line above is not accurate enough anymore:
`_required_safept->push(n);  // save the one closest to the tail`

For one: could there not be multiple such SafePoints? If so: what does it mean to take "the closest"? And we may have multiple backedges / tails, now that we allow irreducible loops.

Comment (eme64) [reply src/hotspot/share/opto/loopnode.cpp:3515]: Other than that the fix looks reasonable to me, thanks for the fix @rwestrel !

**Comment by rwestrel** [comment]
@eme64 thanks for reviewing this (and making suggestions). Does the updated change look ok?

**Review by eme64 — APPROVED** [review]
Thanks for the changes @rwestrel , looks good now!

**Comment by rwestrel** [comment]
@TobiHartmann @eme64 thanks for the reviews.

**Pull request closed without merging**